// @defiyield-app/sdk/guards.cjs.js

var ChainDetailed_possibleTypes = ['ChainDetailed']
module.exports.isChainDetailed = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isChainDetailed"')
  return ChainDetailed_possibleTypes.includes(obj.__typename)
}



var Chain_possibleTypes = ['Chain']
module.exports.isChain = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isChain"')
  return Chain_possibleTypes.includes(obj.__typename)
}



var AssetCategory_possibleTypes = ['AssetCategory']
module.exports.isAssetCategory = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAssetCategory"')
  return AssetCategory_possibleTypes.includes(obj.__typename)
}



var AssetUnderlying_possibleTypes = ['AssetUnderlying']
module.exports.isAssetUnderlying = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAssetUnderlying"')
  return AssetUnderlying_possibleTypes.includes(obj.__typename)
}



var Asset_possibleTypes = ['Asset']
module.exports.isAsset = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAsset"')
  return Asset_possibleTypes.includes(obj.__typename)
}



var AssetBalance_possibleTypes = ['AssetBalance']
module.exports.isAssetBalance = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAssetBalance"')
  return AssetBalance_possibleTypes.includes(obj.__typename)
}



var AssetBalancesResponse_possibleTypes = ['AssetBalancesResponse']
module.exports.isAssetBalancesResponse = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAssetBalancesResponse"')
  return AssetBalancesResponse_possibleTypes.includes(obj.__typename)
}



var AssetPrice_possibleTypes = ['AssetPrice']
module.exports.isAssetPrice = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isAssetPrice"')
  return AssetPrice_possibleTypes.includes(obj.__typename)
}



var Credits_possibleTypes = ['Credits']
module.exports.isCredits = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isCredits"')
  return Credits_possibleTypes.includes(obj.__typename)
}



var ProtocolBalanceToken_possibleTypes = ['ProtocolBalanceToken']
module.exports.isProtocolBalanceToken = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolBalanceToken"')
  return ProtocolBalanceToken_possibleTypes.includes(obj.__typename)
}



var ProtocolBalancePositionEntry_possibleTypes = ['ProtocolBalancePositionEntry']
module.exports.isProtocolBalancePositionEntry = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolBalancePositionEntry"')
  return ProtocolBalancePositionEntry_possibleTypes.includes(obj.__typename)
}



var ProtocolBalancePosition_possibleTypes = ['ProtocolBalancePosition']
module.exports.isProtocolBalancePosition = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolBalancePosition"')
  return ProtocolBalancePosition_possibleTypes.includes(obj.__typename)
}



var ProtocolBalanceChain_possibleTypes = ['ProtocolBalanceChain']
module.exports.isProtocolBalanceChain = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolBalanceChain"')
  return ProtocolBalanceChain_possibleTypes.includes(obj.__typename)
}



var ProtocolFeature_possibleTypes = ['ProtocolFeature']
module.exports.isProtocolFeature = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolFeature"')
  return ProtocolFeature_possibleTypes.includes(obj.__typename)
}



var Protocol_possibleTypes = ['Protocol']
module.exports.isProtocol = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocol"')
  return Protocol_possibleTypes.includes(obj.__typename)
}



var ProtocolBalance_possibleTypes = ['ProtocolBalance']
module.exports.isProtocolBalance = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isProtocolBalance"')
  return ProtocolBalance_possibleTypes.includes(obj.__typename)
}



var RektToken_possibleTypes = ['RektToken']
module.exports.isRektToken = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isRektToken"')
  return RektToken_possibleTypes.includes(obj.__typename)
}



var Rekt_possibleTypes = ['Rekt']
module.exports.isRekt = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isRekt"')
  return Rekt_possibleTypes.includes(obj.__typename)
}



var ShieldIssue_possibleTypes = ['ShieldIssue']
module.exports.isShieldIssue = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isShieldIssue"')
  return ShieldIssue_possibleTypes.includes(obj.__typename)
}



var ShieldTag_possibleTypes = ['ShieldTag']
module.exports.isShieldTag = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isShieldTag"')
  return ShieldTag_possibleTypes.includes(obj.__typename)
}



var ShieldToken_possibleTypes = ['ShieldToken']
module.exports.isShieldToken = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isShieldToken"')
  return ShieldToken_possibleTypes.includes(obj.__typename)
}



var ShieldVersion_possibleTypes = ['ShieldVersion']
module.exports.isShieldVersion = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isShieldVersion"')
  return ShieldVersion_possibleTypes.includes(obj.__typename)
}



var Shield_possibleTypes = ['Shield']
module.exports.isShield = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isShield"')
  return Shield_possibleTypes.includes(obj.__typename)
}



var Query_possibleTypes = ['Query']
module.exports.isQuery = function(obj) {
  if (!obj || !obj.__typename) throw new Error('__typename is missing in "isQuery"')
  return Query_possibleTypes.includes(obj.__typename)
}
