// @defiyield-app/sdk/schema.ts
export type Scalars = {
    Int: number,
    String: string,
    Float: number,
    DateTime: any,
    Boolean: boolean,
}


/** Chain detailed info */
export interface ChainDetailed {
    id: Scalars['Int']
    absoluteChainId: Scalars['String']
    abbr: Scalars['String']
    name: Scalars['String']
    type: Scalars['String']
    __typename: 'ChainDetailed'
}


/** Chain */
export interface Chain {
    id: Scalars['Int']
    abbr: Scalars['String']
    name: Scalars['String']
    __typename: 'Chain'
}


/** Asset Category */
export interface AssetCategory {
    name: Scalars['String']
    code: Scalars['String']
    __typename: 'AssetCategory'
}


/** Asset Underlying */
export interface AssetUnderlying {
    id?: Scalars['Int']
    address: Scalars['String']
    chainId: Scalars['Int']
    name?: Scalars['String']
    displayName?: Scalars['String']
    symbol?: Scalars['String']
    icon?: Scalars['String']
    decimals: Scalars['Int']
    categories: AssetCategory[]
    chain: Chain
    price?: Scalars['Float']
    underlying: AssetUnderlying[]
    reserve?: Scalars['String']
    __typename: 'AssetUnderlying'
}

export interface Asset {
    id?: Scalars['Int']
    address: Scalars['String']
    chainId: Scalars['Int']
    name?: Scalars['String']
    displayName?: Scalars['String']
    symbol?: Scalars['String']
    icon?: Scalars['String']
    decimals: Scalars['Int']
    categories: AssetCategory[]
    chain: Chain
    price?: Scalars['Float']
    underlying: AssetUnderlying[]
    __typename: 'Asset'
}


/** Asset Balance */
export interface AssetBalance {
    asset: Asset
    /** Asset balance */
    balance: Scalars['Float']
    /** Asset price in USD */
    price?: Scalars['Float']
    total: Scalars['Float']
    __typename: 'AssetBalance'
}


/** Asset Balance Response */
export interface AssetBalancesResponse {
    total: Scalars['Float']
    assets: AssetBalance[]
    __typename: 'AssetBalancesResponse'
}


/** Asset Price */
export interface AssetPrice {
    id: Scalars['Int']
    address: Scalars['String']
    chainId: Scalars['Int']
    /** Price in USD */
    price?: Scalars['Float']
    __typename: 'AssetPrice'
}


/** Chain detailed info */
export interface Credits {
    /** Cost for the current call */
    cost: Scalars['Int']
    /** Remaining number of credits */
    remaining: Scalars['Int']
    /** Time at which the current limit resets, a signed integer which represents date and time as number of milliseconds from start of UNIX epoch */
    resetAt?: Scalars['DateTime']
    __typename: 'Credits'
}


/** Protocol balance token info */
export interface ProtocolBalanceToken {
    id?: Scalars['Int']
    address: Scalars['String']
    chainId: Scalars['Int']
    name?: Scalars['String']
    displayName?: Scalars['String']
    symbol?: Scalars['String']
    icon?: Scalars['String']
    decimals: Scalars['Int']
    price?: Scalars['Float']
    /** Token position balance */
    balance?: Scalars['Float']
    /** Position value in USD */
    value?: Scalars['Float']
    /** Token amount */
    amount?: Scalars['Float']
    underlying: ProtocolBalanceToken[]
    __typename: 'ProtocolBalanceToken'
}


/** Protocol balance position entry */
export interface ProtocolBalancePositionEntry {
    /** Year APR */
    apr?: Scalars['Float']
    /** Token amount */
    amount?: Scalars['Float']
    /** Position value in USD */
    value?: Scalars['Float']
    tvl?: Scalars['Float']
    token: ProtocolBalanceToken
    __typename: 'ProtocolBalancePositionEntry'
}


/** Protocol balance position on chain */
export interface ProtocolBalancePosition {
    id: Scalars['String']
    feature: ProtocolFeatureEnum
    borrowed?: ProtocolBalancePositionEntry[]
    supplied?: ProtocolBalancePositionEntry[]
    rewarded?: ProtocolBalancePositionEntry[]
    __typename: 'ProtocolBalancePosition'
}

export type ProtocolFeatureEnum = 'pools' | 'staking' | 'claimable' | 'lending' | 'delegation' | 'lockedBalances' | 'leverageFarming' | 'insurance'


/** Protocol balance on chain */
export interface ProtocolBalanceChain {
    chain: Chain
    featureList: ProtocolFeatureEnum[]
    total: Scalars['Float']
    positions: ProtocolBalancePosition[]
    __typename: 'ProtocolBalanceChain'
}


/** Protocol feature */
export interface ProtocolFeature {
    list: ProtocolFeatureEnum[]
    chain: Chain
    __typename: 'ProtocolFeature'
}


/** Protocol */
export interface Protocol {
    name: Scalars['String']
    slug: Scalars['String']
    features: ProtocolFeature[]
    __typename: 'Protocol'
}


/** Protocol balance */
export interface ProtocolBalance {
    protocol: Protocol
    address: Scalars['String']
    total: Scalars['Float']
    chains: ProtocolBalanceChain[]
    __typename: 'ProtocolBalance'
}


/** Rekt token info */
export interface RektToken {
    name?: Scalars['String']
    addresses: Scalars['String'][]
    __typename: 'RektToken'
}


/** Rekt info model */
export interface Rekt {
    id: Scalars['Int']
    projectName: Scalars['String']
    /** Incident description with HTML tags */
    description?: Scalars['String']
    /** Date in: YYYY-mm-dd */
    date?: Scalars['String']
    fundsLost?: Scalars['String']
    fundsReturned?: Scalars['String']
    chaindIds: Scalars['Int'][]
    category?: Scalars['String']
    issueType?: Scalars['String']
    token: RektToken
    __typename: 'Rekt'
}


/** Shield issue */
export interface ShieldIssue {
    id?: Scalars['Int']
    registryId?: Scalars['String']
    impact?: Scalars['String']
    title?: Scalars['String']
    description?: Scalars['String']
    category?: Scalars['String']
    data?: Scalars['String']
    __typename: 'ShieldIssue'
}


/** Shield tag */
export interface ShieldTag {
    contract?: Scalars['String']
    deployer?: Scalars['String']
    __typename: 'ShieldTag'
}


/** Shield token */
export interface ShieldToken {
    name?: Scalars['String']
    symbol?: Scalars['String']
    decimals: Scalars['Int']
    __typename: 'ShieldToken'
}


/** Shield version */
export interface ShieldVersion {
    scanner?: Scalars['String']
    scanned?: Scalars['String']
    __typename: 'ShieldVersion'
}


/** Shield info */
export interface Shield {
    id: Scalars['Int']
    address?: Scalars['String']
    network?: Scalars['Int']
    name?: Scalars['String']
    logo?: Scalars['String']
    inProgress?: Scalars['Boolean']
    whitelisted?: Scalars['Boolean']
    version?: ShieldVersion
    tags?: ShieldTag
    token?: ShieldToken
    issues?: ShieldIssue[]
    __typename: 'Shield'
}

export interface Query {
    chains: ChainDetailed[]
    assetBalances: AssetBalancesResponse
    assetsPrices: AssetPrice[]
    assets: Asset[]
    assetsCached: Asset[]
    credits: Credits
    protocolBalance: ProtocolBalance[]
    protocols: Protocol[]
    rekts: Rekt[]
    shields: Shield[]
    __typename: 'Query'
}

export type OrderByFields = 'symbol' | 'address' | 'name' | 'chainId' | 'decimals'

export type OrderDirection = 'asc' | 'desc'

export type Sort = 'asc' | 'desc'


/** Chain detailed info */
export interface ChainDetailedRequest{
    id?: boolean | number
    absoluteChainId?: boolean | number
    abbr?: boolean | number
    name?: boolean | number
    type?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Chain */
export interface ChainRequest{
    id?: boolean | number
    abbr?: boolean | number
    name?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Asset Category */
export interface AssetCategoryRequest{
    name?: boolean | number
    code?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Asset Underlying */
export interface AssetUnderlyingRequest{
    id?: boolean | number
    address?: boolean | number
    chainId?: boolean | number
    name?: boolean | number
    displayName?: boolean | number
    symbol?: boolean | number
    icon?: boolean | number
    decimals?: boolean | number
    categories?: AssetCategoryRequest
    chain?: ChainRequest
    price?: boolean | number
    underlying?: AssetUnderlyingRequest
    reserve?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}

export interface AssetRequest{
    id?: boolean | number
    address?: boolean | number
    chainId?: boolean | number
    name?: boolean | number
    displayName?: boolean | number
    symbol?: boolean | number
    icon?: boolean | number
    decimals?: boolean | number
    categories?: AssetCategoryRequest
    chain?: ChainRequest
    price?: boolean | number
    underlying?: AssetUnderlyingRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Asset Balance */
export interface AssetBalanceRequest{
    asset?: AssetRequest
    /** Asset balance */
    balance?: boolean | number
    /** Asset price in USD */
    price?: boolean | number
    total?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Asset Balance Response */
export interface AssetBalancesResponseRequest{
    total?: boolean | number
    assets?: AssetBalanceRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Asset Price */
export interface AssetPriceRequest{
    id?: boolean | number
    address?: boolean | number
    chainId?: boolean | number
    /** Price in USD */
    price?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Chain detailed info */
export interface CreditsRequest{
    /** Cost for the current call */
    cost?: boolean | number
    /** Remaining number of credits */
    remaining?: boolean | number
    /** Time at which the current limit resets, a signed integer which represents date and time as number of milliseconds from start of UNIX epoch */
    resetAt?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol balance token info */
export interface ProtocolBalanceTokenRequest{
    id?: boolean | number
    address?: boolean | number
    chainId?: boolean | number
    name?: boolean | number
    displayName?: boolean | number
    symbol?: boolean | number
    icon?: boolean | number
    decimals?: boolean | number
    price?: boolean | number
    /** Token position balance */
    balance?: boolean | number
    /** Position value in USD */
    value?: boolean | number
    /** Token amount */
    amount?: boolean | number
    underlying?: ProtocolBalanceTokenRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol balance position entry */
export interface ProtocolBalancePositionEntryRequest{
    /** Year APR */
    apr?: boolean | number
    /** Token amount */
    amount?: boolean | number
    /** Position value in USD */
    value?: boolean | number
    tvl?: boolean | number
    token?: ProtocolBalanceTokenRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol balance position on chain */
export interface ProtocolBalancePositionRequest{
    id?: boolean | number
    feature?: boolean | number
    borrowed?: ProtocolBalancePositionEntryRequest
    supplied?: ProtocolBalancePositionEntryRequest
    rewarded?: ProtocolBalancePositionEntryRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol balance on chain */
export interface ProtocolBalanceChainRequest{
    chain?: ChainRequest
    featureList?: boolean | number
    total?: boolean | number
    positions?: ProtocolBalancePositionRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol feature */
export interface ProtocolFeatureRequest{
    list?: boolean | number
    chain?: ChainRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol */
export interface ProtocolRequest{
    name?: boolean | number
    slug?: boolean | number
    features?: ProtocolFeatureRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Protocol balance */
export interface ProtocolBalanceRequest{
    protocol?: ProtocolRequest
    address?: boolean | number
    total?: boolean | number
    chains?: ProtocolBalanceChainRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Rekt token info */
export interface RektTokenRequest{
    name?: boolean | number
    addresses?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Rekt info model */
export interface RektRequest{
    id?: boolean | number
    projectName?: boolean | number
    /** Incident description with HTML tags */
    description?: boolean | number
    /** Date in: YYYY-mm-dd */
    date?: boolean | number
    fundsLost?: boolean | number
    fundsReturned?: boolean | number
    chaindIds?: boolean | number
    category?: boolean | number
    issueType?: boolean | number
    token?: RektTokenRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Shield issue */
export interface ShieldIssueRequest{
    id?: boolean | number
    registryId?: boolean | number
    impact?: boolean | number
    title?: boolean | number
    description?: boolean | number
    category?: boolean | number
    data?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Shield tag */
export interface ShieldTagRequest{
    contract?: boolean | number
    deployer?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Shield token */
export interface ShieldTokenRequest{
    name?: boolean | number
    symbol?: boolean | number
    decimals?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Shield version */
export interface ShieldVersionRequest{
    scanner?: boolean | number
    scanned?: boolean | number
    __typename?: boolean | number
    __scalar?: boolean | number
}


/** Shield info */
export interface ShieldRequest{
    id?: boolean | number
    address?: boolean | number
    network?: boolean | number
    name?: boolean | number
    logo?: boolean | number
    inProgress?: boolean | number
    whitelisted?: boolean | number
    version?: ShieldVersionRequest
    tags?: ShieldTagRequest
    token?: ShieldTokenRequest
    issues?: ShieldIssueRequest
    __typename?: boolean | number
    __scalar?: boolean | number
}

export interface QueryRequest{
    chains?: ChainDetailedRequest
    assetBalances?: [{chainId: Scalars['Int'],walletAddress: Scalars['String']},AssetBalancesResponseRequest]
    assetsPrices?: [{assets: AssetRequest[]},AssetPriceRequest]
    assets?: [{first?: (Scalars['Int'] | null),skip?: (Scalars['Int'] | null),where?: (AssetFilterInput | null),orderBy?: (OrderByFields | null),orderDirection?: (OrderDirection | null)},AssetRequest] | AssetRequest
    assetsCached?: [{first?: (Scalars['Int'] | null),skip?: (Scalars['Int'] | null),where?: (AssetFilterInput | null),orderBy?: (OrderByFields | null),orderDirection?: (OrderDirection | null)},AssetRequest] | AssetRequest
    credits?: CreditsRequest
    protocolBalance?: [{balances: ProtocolBalanceRequest[]},ProtocolBalanceRequest]
    protocols?: [{name?: (Scalars['String'] | null),slug?: (Scalars['String'] | null),chainIds?: (Scalars['Int'][] | null),features?: (ProtocolFeatureEnum[] | null)},ProtocolRequest] | ProtocolRequest
    rekts?: [{pageNumber?: (Scalars['Int'] | null),
    /** Number of records per page */
    pageSize?: (Scalars['Int'] | null),
    /** Search by project name, token name, token address */
    searchText?: (Scalars['String'] | null),chainIds?: (Scalars['Int'][] | null),orderBy?: (GetRektsOrder | null)},RektRequest] | RektRequest
    shields?: [{addresses: Scalars['String'][],chainId: Scalars['Int']},ShieldRequest]
    __typename?: boolean | number
    __scalar?: boolean | number
}

export interface AssetRequest {chainId: Scalars['Int'],address: Scalars['String']}

export interface AssetFilterInput {chainId?: (Scalars['Int'] | null),addresses?: (Scalars['String'][] | null),symbols?: (Scalars['String'][] | null)}

export interface ProtocolBalanceRequest {chainIds: Scalars['Int'][],walletAddress: Scalars['String'],protocolName: Scalars['String']}

export interface GetRektsOrder {project?: (Sort | null),ticker?: (Sort | null),fundsLost?: (Sort | null),chain?: (Sort | null),issue?: (Sort | null),date?: (Sort | null),category?: (Sort | null)}


const ChainDetailed_possibleTypes: string[] = ['ChainDetailed']
export const isChainDetailed = (obj?: { __typename?: any } | null): obj is ChainDetailed => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isChainDetailed"')
  return ChainDetailed_possibleTypes.includes(obj.__typename)
}



const Chain_possibleTypes: string[] = ['Chain']
export const isChain = (obj?: { __typename?: any } | null): obj is Chain => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isChain"')
  return Chain_possibleTypes.includes(obj.__typename)
}



const AssetCategory_possibleTypes: string[] = ['AssetCategory']
export const isAssetCategory = (obj?: { __typename?: any } | null): obj is AssetCategory => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAssetCategory"')
  return AssetCategory_possibleTypes.includes(obj.__typename)
}



const AssetUnderlying_possibleTypes: string[] = ['AssetUnderlying']
export const isAssetUnderlying = (obj?: { __typename?: any } | null): obj is AssetUnderlying => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAssetUnderlying"')
  return AssetUnderlying_possibleTypes.includes(obj.__typename)
}



const Asset_possibleTypes: string[] = ['Asset']
export const isAsset = (obj?: { __typename?: any } | null): obj is Asset => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAsset"')
  return Asset_possibleTypes.includes(obj.__typename)
}



const AssetBalance_possibleTypes: string[] = ['AssetBalance']
export const isAssetBalance = (obj?: { __typename?: any } | null): obj is AssetBalance => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAssetBalance"')
  return AssetBalance_possibleTypes.includes(obj.__typename)
}



const AssetBalancesResponse_possibleTypes: string[] = ['AssetBalancesResponse']
export const isAssetBalancesResponse = (obj?: { __typename?: any } | null): obj is AssetBalancesResponse => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAssetBalancesResponse"')
  return AssetBalancesResponse_possibleTypes.includes(obj.__typename)
}



const AssetPrice_possibleTypes: string[] = ['AssetPrice']
export const isAssetPrice = (obj?: { __typename?: any } | null): obj is AssetPrice => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isAssetPrice"')
  return AssetPrice_possibleTypes.includes(obj.__typename)
}



const Credits_possibleTypes: string[] = ['Credits']
export const isCredits = (obj?: { __typename?: any } | null): obj is Credits => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isCredits"')
  return Credits_possibleTypes.includes(obj.__typename)
}



const ProtocolBalanceToken_possibleTypes: string[] = ['ProtocolBalanceToken']
export const isProtocolBalanceToken = (obj?: { __typename?: any } | null): obj is ProtocolBalanceToken => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolBalanceToken"')
  return ProtocolBalanceToken_possibleTypes.includes(obj.__typename)
}



const ProtocolBalancePositionEntry_possibleTypes: string[] = ['ProtocolBalancePositionEntry']
export const isProtocolBalancePositionEntry = (obj?: { __typename?: any } | null): obj is ProtocolBalancePositionEntry => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolBalancePositionEntry"')
  return ProtocolBalancePositionEntry_possibleTypes.includes(obj.__typename)
}



const ProtocolBalancePosition_possibleTypes: string[] = ['ProtocolBalancePosition']
export const isProtocolBalancePosition = (obj?: { __typename?: any } | null): obj is ProtocolBalancePosition => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolBalancePosition"')
  return ProtocolBalancePosition_possibleTypes.includes(obj.__typename)
}



const ProtocolBalanceChain_possibleTypes: string[] = ['ProtocolBalanceChain']
export const isProtocolBalanceChain = (obj?: { __typename?: any } | null): obj is ProtocolBalanceChain => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolBalanceChain"')
  return ProtocolBalanceChain_possibleTypes.includes(obj.__typename)
}



const ProtocolFeature_possibleTypes: string[] = ['ProtocolFeature']
export const isProtocolFeature = (obj?: { __typename?: any } | null): obj is ProtocolFeature => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolFeature"')
  return ProtocolFeature_possibleTypes.includes(obj.__typename)
}



const Protocol_possibleTypes: string[] = ['Protocol']
export const isProtocol = (obj?: { __typename?: any } | null): obj is Protocol => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocol"')
  return Protocol_possibleTypes.includes(obj.__typename)
}



const ProtocolBalance_possibleTypes: string[] = ['ProtocolBalance']
export const isProtocolBalance = (obj?: { __typename?: any } | null): obj is ProtocolBalance => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isProtocolBalance"')
  return ProtocolBalance_possibleTypes.includes(obj.__typename)
}



const RektToken_possibleTypes: string[] = ['RektToken']
export const isRektToken = (obj?: { __typename?: any } | null): obj is RektToken => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isRektToken"')
  return RektToken_possibleTypes.includes(obj.__typename)
}



const Rekt_possibleTypes: string[] = ['Rekt']
export const isRekt = (obj?: { __typename?: any } | null): obj is Rekt => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isRekt"')
  return Rekt_possibleTypes.includes(obj.__typename)
}



const ShieldIssue_possibleTypes: string[] = ['ShieldIssue']
export const isShieldIssue = (obj?: { __typename?: any } | null): obj is ShieldIssue => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isShieldIssue"')
  return ShieldIssue_possibleTypes.includes(obj.__typename)
}



const ShieldTag_possibleTypes: string[] = ['ShieldTag']
export const isShieldTag = (obj?: { __typename?: any } | null): obj is ShieldTag => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isShieldTag"')
  return ShieldTag_possibleTypes.includes(obj.__typename)
}



const ShieldToken_possibleTypes: string[] = ['ShieldToken']
export const isShieldToken = (obj?: { __typename?: any } | null): obj is ShieldToken => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isShieldToken"')
  return ShieldToken_possibleTypes.includes(obj.__typename)
}



const ShieldVersion_possibleTypes: string[] = ['ShieldVersion']
export const isShieldVersion = (obj?: { __typename?: any } | null): obj is ShieldVersion => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isShieldVersion"')
  return ShieldVersion_possibleTypes.includes(obj.__typename)
}



const Shield_possibleTypes: string[] = ['Shield']
export const isShield = (obj?: { __typename?: any } | null): obj is Shield => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isShield"')
  return Shield_possibleTypes.includes(obj.__typename)
}



const Query_possibleTypes: string[] = ['Query']
export const isQuery = (obj?: { __typename?: any } | null): obj is Query => {
  if (!obj?.__typename) throw new Error('__typename is missing in "isQuery"')
  return Query_possibleTypes.includes(obj.__typename)
}
